VAULT-9863 Kubernetes Overview Page

kiannaquach opened the pull request:
Adds overview page for Kubernetes secrets engine
[Overview Figma Design here](https://www.figma.com/file/rB5gYCA95F9aKMc7I2Qx0d/Kubernetes-Secret-Engine-Handoff?node-id=1%3A719&t=LEZsdtAweDMP5xoW-0)

**Overview when Kubernetes is not configured:**
<img width="1728" alt="Screen Shot 2022-12-05 at 12 47 22 PM" src="https://user-images.githubusercontent.com/30884335/205741713-fd22787b-67d9-479d-92b1-de8ecf025791.png">

**Overview when Kubernetes has no roles:**
<img width="1728" alt="Screen Shot 2022-12-05 at 12 47 38 PM" src="https://user-images.githubusercontent.com/30884335/205741712-595acfb5-f0a4-46c0-9456-4552bdd64662.png">

**Overview when there are roles**

https://user-images.githubusercontent.com/30884335/205747932-015021b7-3891-4ff5-a9ff-e8257bc09205.mov

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault?pullRequest=18232)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault?pullRequest=18232)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**Kianna Quach** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA. If you already have a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault?pullRequest=18232).</sub>

zofskeez commented on ui/lib/kubernetes/addon/components/page/overview.hbs:
```diff
@@ -0,0 +1,53 @@
+        </div>
+        <p>The number of Vault roles being used to generate Kubernetes credentials.</p>
+        <h2 class="title has-font-weight-normal is-3 has-top-margin-l">
+          {{if @model.roles.length @model.roles.length "None"}}
```
I think you could use the `or` helper here instead
```suggestion
          {{or @model.roles.length "None"}}
```

zofskeez commented on ui/lib/kubernetes/addon/components/page/overview.js:
```diff
@@ -0,0 +1,35 @@
+
+  constructor() {
+    super(...arguments);
+    this.roles = this.args.model.roles.toArray();
```
You should be able to remove the tracked `roles` property and access this directly in the template via `@model.roles`.

zofskeez commented on ui/lib/kubernetes/addon/components/page/overview.js:
```diff
@@ -0,0 +1,35 @@
+    this.roles = this.args.model.roles.toArray();
+  }
+
+  get isDisabled() {
+    return !this.selectedRole;
+  }
```
I'm wondering since this is a simple expression if it would be better to use `disabled={{not this.selectedRole}}` directly in the template?

zofskeez commented on ui/lib/kubernetes/addon/components/page/overview.js:
```diff
@@ -0,0 +1,35 @@
+
+  @action
+  generateCredential() {
+    const { selectedRole } = this;
```
Rather than assigning `selectedRole` to a `const` it might be cleaner to access it directly via `this.selectedRole`.

zofskeez commented on ui/tests/integration/components/kubernetes/page/overview-test.js:
```diff
@@ -0,0 +1,97 @@
+        type: 'kubernetes',
+      },
+    });
+    this.store.pushPayload('kubernetes/config', {
+      modelName: 'kubernetes/config',
+      backend: 'kubernetes-test',
+      ...this.server.create('kubernetes-config'),
+    });
+    this.store.pushPayload('kubernetes/role', {
+      modelName: 'kubernetes/role',
+      backend: 'kubernetes-test',
+      ...this.server.create('kubernetes-role'),
+    });
+    this.store.pushPayload('kubernetes/role', {
+      modelName: 'kubernetes/role',
+      backend: 'kubernetes-test',
+      ...this.server.create('kubernetes-role'),
+    });
```
You can use the kubernetes mirage handler if you would like by setting and unsetting it in the before and after hooks:
```js
hooks.before(function () {
  ENV['ember-cli-mirage'].handler = 'kubernetes';
});

hooks.after(function () {
  ENV['ember-cli-mirage'].handler = null;
});
```
This is more of an FYI and what you did here is perfectly fine. The default scenario is not executed in the test so you would have to seed the DB initially in the before hook:
```js
this.server.create('kubernetes-config');
this.server.createList('kubernetes-role', 2);
```
Totally up to you if you would like to explore that route but I just wanted to let you know that option exists for the future.

zofskeez replied:
Looking at this again today with a clear head I realize now that this component is not responsible for loading the models so I led you astray with this suggestion :disappointed:. Your original approach was best! Sorry for the run around. Keep this in mind for acceptance tests when the endpoints are getting hit.

zofskeez commented on ui/tests/integration/components/kubernetes/page/overview-test.js:
```diff
@@ -0,0 +1,97 @@
+    };
+  });
+
+  test('it should display role card', async function (assert) {
```
It would be nice to also add a test if there is no config to ensure that it shows the call to action.

zofskeez reviewed (changes requested):
Great work and very impressive for your first PR :tada:! I left a few comments for some possible minor changes but well done :clap:

kiannaquach commented:
Thanks for the feedback, @zofskeez!! I updated the tests so that it is using the Kubernetes mirage handler and also removed some extra stuff in the `component.js` file in favor of template helpers!

zofskeez commented on ui/lib/kubernetes/addon/components/page/overview.js:
```diff
@@ -6,30 +6,21 @@ import { action } from '@ember/object';
-  @tracked roles = [];
   @tracked selectedRole = null;
-  @tracked roleOptions = this.roles.map((role) => {
+  @tracked roleOptions = this.args.model.roles.map((role) => {
```
Sorry I didn't notice this the first time around but I think it's probably better to set `roleOptions` in the constructor after calling `super`.

kiannaquach replied:
No problem! Just fixed!

zofskeez reviewed (approved):
:tada:
